### PR TITLE
add abstract name resolver factory

### PR DESCRIPTION
**Pull request by AdoHe**
@lburgazzoli ptal.

### CODE REVIEW

**Comment by AdoHe** [comment]
@lburgazzoli I just add init discovery service based name resolver, still wip. ptal

**Comment by lburgazzoli** [comment]
Any chance to get it merged, maybe in a different branch ?

File: src/main/java/com/coreos/jetcd/resolver/DiscoveryEtcdNameResolver.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import java.net.URI;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * A discovery service based {@link NameResolver}. DiscoveryEtcdNameResolver will
+ * query configured discovery URL periodically to get endpoints addresses.
+ */
+public class DiscoveryEtcdNameResolver extends NameResolver {
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/resolver/DiscoveryEtcdNameResolver.java]: if you refer to public discovery.etcd.io, it does not really work. discovery url only expose peer url, not client url.

Comment (xiang90) [inline src/main/java/com/coreos/jetcd/resolver/DiscoveryEtcdNameResolver.java]: i think the first naming resolver would be a dns srv resolver with client prefix. see https://github.com/coreos/etcd/blob/master/Documentation/op-guide/clustering.md#dns-discovery

**Comment by xiang90** [comment]
@lburgazzoli @AdoHe I think we need to have srv discover first. the discover token one does not really work today. see my inline comments.

File: src/main/java/com/coreos/jetcd/resolver/DiscoveryEtcdNameResolver.java (hunk 1 above)
Comment (AdoHe) [inline src/main/java/com/coreos/jetcd/resolver/DiscoveryEtcdNameResolver.java]: Yes, I am wondering whether we should add a discovery service based name resolver.

**Comment by AdoHe** [comment]
@xiang90 so perhaps we just need two name resolvers: `simple` and `dns+srv`.

**Comment by xiang90** [comment]
@AdoHe What does simple do? nothing but return the original endpoints?

**Comment by AdoHe** [comment]
@xiang90 Yes, simple name resolver just write the original endpoints to listener.

**Comment by xiang90** [comment]
sure. thanks.

**Comment by AdoHe** [comment]
@lburgazzoli I will update this today and merge it so you can start the srv based name resolver.

**Comment by xiang90** [comment]
lgtm. can you please squash the commits into one?

**Pull request merged**